lib/Dataflow_analysis.ml

```diff
@@ -0,0 +1,800 @@
+open Core_kernel
+open Mir
+
+(***********************************)
+(* Basic datatypes                 *)
+(***********************************)
+
+(**
+   A label is a unique identifier for a node in the dataflow/dependency graph, and
+   usually corresponds to one node in the Mir.
+*)
+type label = int [@@deriving sexp]
+
+(**
+   A 'reaching dependency' (or reaching_dep or RD) statement (v, l) says that the variable
+   v could have been affected at the label l.
+*)
+type reaching_dep = expr * int [@@deriving sexp, hash, compare]
+
+(**
+   A reaching dependency set holds the set of reaching dependencies that could be true at
+   some point.
+*)
+module ReachingDepSet = Set.Make (struct
+  type t = reaching_dep
+
+  let compare : reaching_dep -> reaching_dep -> int = compare
+  let sexp_of_t = sexp_of_reaching_dep
+  let t_of_sexp = reaching_dep_of_sexp
+end)
+
+module LabelMap = Map.Make (struct
+  type t = label
+
+  let compare : int -> int -> int = compare
+  let sexp_of_t = sexp_of_int
+  let t_of_sexp = int_of_sexp
+end)
+
+module LabelSet = Set.Make (struct
+  type t = label
+
+  let compare : int -> int -> int = compare
+  let sexp_of_t = sexp_of_int
+  let t_of_sexp = int_of_sexp
+end)
+
+module ExprSet = Set.Make (struct
+  type t = expr
+
+  let compare : expr -> expr -> int = compare
+  let sexp_of_t = sexp_of_expr
+  let t_of_sexp = expr_of_sexp
+end)
+
+(**
+   Description of where a node in the dependency graph came from, where MirNode is the
+   location from an Mir.loc_stmt
+ *)
+type source_loc =
+  | MirNode of string
+  | StartOfBlock
+  | TargetTerm of {term: expr; assignment_label: label}
+[@@deriving sexp]
+
+(**
+   Information to be collected about each node
+   * dep_sets: Information about how the label effects the dependency set
+   * possible_previous: The set of nodes that could have immediately preceded this node
+     under some execution of the program
+   * rhs_set: The 'right hand side' set of variables that affect the value or behavior of
+     this node
```

Comment:
That's right about rhs_set.

Again I regret not first writing out the outline of the approach. Basically:
1. Traverse the MIR to build up node information including RD update functions,
2. Find fixed-point solutions of the RD update functions,
3. Use the node and RD information to complete various useful analysis.

So each of the other fields are useful for the dependency analysis at the end. Specifically, dep_sets and possible_previous let us find the data dependencies and controlflow lets us find the control dependencies.

Reply:
Hmm cool. Let's sit down sometime and talk about how we intend to use this module and the design here - I was thinking about just keeping track of which Vars a node needs and which ones it affects (modulo indices), I guess because I was thinking about a few specific simple optimization passes, but I don't know much about the full extent of what you might need for passes I haven't read about in the literature.

Reply:
It would be great to sit down and discuss. I'm quite flexible this week. Maybe Wednesday afternoon or any time Thursday?

Reply:
Let's meet on Thursday afternoon! I have to go in that day anyway :) Maybe around 2pm?

Reply:
2pm Thursday sounds great. Is your office de-molded?

Reply:
Yes, but my desk is gone. But let's try anyway.

Reply:
I'm excited to talk this over in person - I have been saying it's possible to do interesting optimizations in the Stan compiler for a long time and you're building the first directly critical parts of that infrastructure :D